raiden/raiden_event_handler.py

```diff
@@ -197,22 +203,24 @@ def handle_paymentsentsuccess(
             raiden: RaidenService,
             payment_sent_success_event: EventPaymentSentSuccess,
     ):
-        assert payment_sent_success_event.identifier in raiden.identifier_to_results
-
-        result = raiden.identifier_to_results[payment_sent_success_event.identifier]
-        result.set(True)
-        del raiden.identifier_to_results[payment_sent_success_event.identifier]
+        payment_status = raiden.identifiers_to_statuses.pop(
+            payment_sent_success_event.identifier,
+            None,
+        )
+        assert payment_status is not None
+        payment_status.payment_done.set(True)
 
     def handle_paymentsentfailed(
             self,
             raiden: RaidenService,
             payment_sent_failed_event: EventPaymentSentFailed,
     ):
-        assert payment_sent_failed_event.identifier in raiden.identifier_to_results
-
-        result = raiden.identifier_to_results[payment_sent_failed_event.identifier]
-        result.set(False)
-        del raiden.identifier_to_results[payment_sent_failed_event.identifier]
+        payment_status = raiden.identifiers_to_statuses.pop(
+            payment_sent_failed_event.identifier,
+            None,
+        )
+        assert payment_status is not None
```

Comment:
Is there any way we can get here two times so that the `payment_status` is already popped? This assert scares me a bit.

Reply:
this assert is equivalent to what was there before:

```python
assert payment_sent_failed_event.identifier in raiden.identifier_to_results
```

Discussion about the mapping where done here:

https://github.com/raiden-network/raiden/issues/2159

We rely on having only one payment in transit with a given id at any point in time:

https://github.com/raiden-network/raiden/blob/bf61915d3b5549348eb7b7f50ebebca973a5a022/raiden/raiden_service.py#L692-L693

This is what we use to avoid making the same payment twice on retries in the REST API: 

https://github.com/raiden-network/raiden/blob/bf61915d3b5549348eb7b7f50ebebca973a5a022/raiden/api/rest.py#L953-L959

I'm fine with removing the assert, but what are the semantics of the endpoint payment endpoint that you have in mind?

Reply:
fine by me. So if this assert hits its a programming error, so a bug.

Reply:
What about if an `EventPaymentSentFailed` is emitted without a `identifiers_to_statuses` present? e.g. I can see this event can get emitted in a few conditions on `handle_send_directtransfer`, including the channel not being open, so if the user/client makes a mistake and tries to direct transfer to a not-open channel, a `EventPaymentSentFailed` will be emitted without a identifier on this dict, which will trigger the assert and crash the node, while it should only become an error code, not be fatal to the node. I know directtransfer isn't a good example as it isn't exposed throught the APIs, but the handlers are still there, it's just an example.
Asserts in the state handlers should be added very carefully, and all possible conditions in which an event may get emitted must be considered, or we'll crash the node.

Reply:
Oh, nvm, I can see the dict is populated right before the `handle_state_change` for direct transfers.. still, the heads-up for asserts here holds.